src/test/codegen/issue-37945.rs

```diff
@@ -1,26 +1,34 @@
-// compile-flags: -O
+// compile-flags: -O -Zmerge-functions=disabled
 // ignore-x86
 // ignore-arm
 // ignore-emscripten
 // ignore-gnux32
 // ignore 32-bit platforms (LLVM has a bug with them)
 
-// See issue #37945.
+// Check that LLVM understands that `Iter` pointer is not null. Issue #37945.
 
 #![crate_type = "lib"]
 
 use std::slice::Iter;
 
-// CHECK-LABEL: @is_empty_1
 #[no_mangle]
 pub fn is_empty_1(xs: Iter<f32>) -> bool {
-// CHECK-NOT: icmp eq float* {{.*}}, null
+// CHECK-LABEL: @is_empty_1(
+// CHECK-NEXT:  start:
+// CHECK-NEXT:    [[A:%.*]] = icmp ne i32* %xs.1, null
```

Comment:
> Check that LLVM understands that `Iter` pointer is not null.

If we have `icmp ne ..., null`, it seems like we're failing the thing we meant to test, no?

I don't think we should try to compare exactly what LLVM IR comes out, but perhaps something like:

```
/// CHECK-NOT: icmp {{.*}}, null
```

Reply:
> If we have `icmp ne ..., null`, it seems like we're failing the thing we meant to test, no?

It was fixed upstream by canonicalizing to assume.

Reply:
By "it was fixed upstream", do you mean that will be in LLVM 12? Is there something we could cherry-pick?

My point is the test meant to check that LLVM knows it's not null, so we had a `CHECK-NOT: icmp eq ..., null` to that effect. Your change reveals that while we _were_ passing that test, we're getting `icmp ne ..., null` instead, which seems just as bad. So it seems like a real bug with or without your change.

Reply:
The fact that pointer is not-null is expressed through `llvm.assume` that follows, see next check.

Reply:
If you mean these `assume`s, then the bug is that they're apparently not as effective as we want:

https://github.com/rust-lang/rust/blob/f5230fbf76bafd86ee4376a0e26e551df8d17fec/library/core/src/slice/iter/macros.rs#L142-L145

Reply:
Hmm, ok, I see what you mean...